### PR TITLE
Avoid dereference-of-casted-pointer

### PR DESCRIPTION
**Pull request by sipa**
And prefer a static_cast to the intended reference type.

### CODE REVIEW

File: src/core_memusage.h [hunk 1]
```diff
@@ -10,7 +10,7 @@
 #include "memusage.h"
 
 static inline size_t RecursiveDynamicUsage(const CScript& script) {
-    return memusage::DynamicUsage(*static_cast<const CScriptBase*>(&script));
+    return memusage::DynamicUsage(static_cast<const CScriptBase&>(script));
```
Comment (promag) [inline src/core_memusage.h]: Remove static cast?
```cpp
  return memusage::DynamicUsage(script);
```
In my system it builds fine.

File: src/wallet/walletdb.cpp [hunk 2]
```diff
@@ -94,23 +94,23 @@ bool CWalletDB::WriteMasterKey(unsigned int nID, const CMasterKey& kMasterKey)
 
 bool CWalletDB::WriteCScript(const uint160& hash, const CScript& redeemScript)
 {
-    return WriteIC(std::make_pair(std::string("cscript"), hash), *(const CScriptBase*)(&redeemScript), false);
+    return WriteIC(std::make_pair(std::string("cscript"), hash), static_cast<const CScriptBase&>(redeemScript), false);
```
Comment (promag) [inline src/wallet/walletdb.cpp]: Removing the cast, shouldn't the compiler resolve to:
```cpp
template<typename Stream, unsigned int N, typename T>
inline void Serialize(Stream& os, const prevector<N, T>& v)
```
If not I thing we could implement `CScript::Serialize` to call super and drop all of these casts.

**Comment by practicalswift** [comment]
Concept ACK 8be4385d64c5a50c63834b75c06d1b27d66ddb62

Regarding the `static_cast` preference, see also PR #10498: "Use static_cast instead of C-style casts for non-fundamental types".

File: src/core_memusage.h (hunk 1 above)
Comment (sipa) [inline src/core_memusage.h]: Nice catch, fixed.

File: src/wallet/walletdb.cpp (hunk 2 above)
Comment (sipa) [inline src/wallet/walletdb.cpp]: I somehow remembered that this didn't work, but it seems I was wrong. Fixed.

**Comment by promag** [comment]
tACK.

**Review by gmaxwell — APPROVED** [review]
utACK

**Comment by TheBlueMatt** [comment]
utACK 0aadc11fd88b298c7af2dfb69763b2c67dc6b7b0

**Review by dcousens — APPROVED** [review]

**Comment by laanwj** [comment]
> And prefer a static_cast to the intended reference type.

What is the rationale here? Just code cleanup?

**Comment by sipa** [comment]
Yes, cleanup. The existing practice risks hiding some bugs.

**Comment by jonasschnelli** [comment]
utACK 0aadc11fd88b298c7af2dfb69763b2c67dc6b7b0

**Pull request merged**